kubernetes/updater/pkg/img/validator.go

```diff
@@ -0,0 +1,97 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package img
+
+import (
+	"context"
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+
+	"github.com/docker/distribution/reference"
+	"github.com/opencontainers/go-digest"
+)
+
+type NamedTaggedDigested interface {
+	reference.Named
+	reference.Tagged
+	reference.Digested
+}
+
+type Validator interface {
+	Name() string
+	ValidateAndResolveDigest(ctx context.Context, image reference.NamedTagged) ([]NamedTaggedDigested, error)
+}
+
+func AtLeastOneValid(ctx context.Context, image reference.NamedTagged, validators ...Validator) (bool, []NamedTaggedDigested) {
+	var resolvedImages []NamedTaggedDigested
+	var err error
+	log := ctrllog.FromContext(ctx).V(1)
+	for _, validator := range validators {
+		resolvedImages, err = validator.ValidateAndResolveDigest(ctx, image)
+		log.Info("Image validation result", "img", image, "validator", validator.Name(), "err", err)
+		// There is no need to continue if we have at least a valid signature
+		// All authorities are expected to sign the same thing
+		if err == nil {
+			log.Info("Image approved by the validator", "img", image, "validator", validator.Name(), "resolvedImages", resolvedImages)
+			return true, resolvedImages
+		}
+	}
+	log.Info("No validator approved the img", "img", image)
```

Comment:
```suggestion
	log.Warn("Unable to validate the image", "img", image)
```

Reply:
This logger is provided by controller-utils and implements the `logr` interface. It does not expose a `Warn()` method. We can play with verbosity levels to work around this limitation.

Reply:
I kept this log as Info() but propagated the information through an error instead.